scipy/linalg/tests/test_lapack.py

```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
+
+    # create the matrix in accordance with the data type
+    du = my_rand(n-1, dtype=dtype)
+    d = my_rand(n, dtype=dtype)
+    dl = my_rand(n-1, dtype=dtype)
+
+    diag_cpy = [dl.copy(), d.copy(), du.copy()]
+
+    A = np.diag(d) + np.diag(dl, -1) + np.diag(du, 1)
+    x = np.random.rand(n)
+    b = A @ x
+
+    gttrf, gttrs = get_lapack_funcs(('gttrf', 'gttrs'), dtype=dtype)
+
+    _dl, _d, _du, du2, ipiv, info = gttrf(dl, d, du)
+    # test to assure that the inputs of ?gttrf are unmodified
+    assert_array_equal(dl, diag_cpy[0])
+    assert_array_equal(d, diag_cpy[1])
+    assert_array_equal(du, diag_cpy[2])
+
+    # generate L and U factors from ?gttrf return values
+    # L/U are lower/upper triangular by construction (initially and at end)
+    U = np.diag(_d, 0) + np.diag(_du, 1) + np.diag(du2, 2)
+    L = np.eye(n, dtype=dtype)
+
+    for i, m in enumerate(_dl):
+        # L is given in a factored form.
+        # See http://www.hpcavf.uclan.ac.uk/softwaredoc/sgi_scsl_html/sgi_html/ch03.html
+        piv = ipiv[i] - 1
+        # right multiply by permutation matrix
+        L[:, [i, piv]] = L[:, [piv, i]]
+        # right multiply by Li, rank-one modification of identity
+        L[:, i] += L[:, i+1]*m
+
+    # one last permutation
+    i, piv = -1, ipiv[-1] - 1
+    # right multiply by final permutation matrix
+    L[:, [i, piv]] = L[:, [piv, i]]
+
+    _A = L @ U
+    # check that the outputs of ?gttrf define an LU decomposition of A
+    assert_allclose(A, _A, atol=rtol)
+
+    b_cpy = b.copy()
+    x_gttrs, info = gttrs(_dl, _d, _du, du2, ipiv, b)
+    # test that the inputs of ?gttrs are unmodified
+    assert_array_equal(b, b_cpy)
+    # test that the result of ?gttrs matches the expected input
+    assert_allclose(x, x_gttrs, rtol=rtol)
+
+    # test that ?gttrf and ?gttrs work with transposal options
+    if dtype in REAL_DTYPES:
+        trans = "T"
+        b_trans = A.T @ x
+    else:
+        trans = "C"
+        b_trans = A.conj().T @ x
+
+    x_gttrs, info = gttrs(_dl, _d, _du, du2, ipiv, b_trans, trans=trans)
+    assert_allclose(x, x_gttrs, rtol=rtol)
+
+    # test that ValueError is raised with incompatible matrix shapes
+    with assert_raises(ValueError):
+        gttrf(dl[:-1], d, du)
+    with assert_raises(ValueError):
+        gttrf(dl, d[:-1], du)
+    with assert_raises(ValueError):
+        gttrf(dl, d, du[:-1])
+
+    # test that matrix of size n=2 raises exception
+    with assert_raises(Exception):
+        gttrf(dl[0], d[:1], du[0])
+
+    # test that singular (row of all zeroes) matrix fails via info
+    du[0] = 0
+    d[0] = 0
+    __dl, __d, __du, _du2, _ipiv, _info = gttrf(dl, d, du)
+    np.testing.assert_(_info != 0,
+                       "?gttrf: singular matrix ret info == 0")
+
+
+def my_rand(shape, dtype):
+    # generates a random matrix of desired data type
+    x = np.random.rand(shape) + np.random.rand(shape)*1.0j
+    return x.astype(dtype)
```

Comment:
On my machine, this raises a ``ComplexWarning`` which would fail any other tests that do not explicitly filter for these. If you want to avoid the warning, maybe try using the dtype as a conditional? e.g.:

```python
if dtype in COMPLEX_DTYPES:
    return (np.random.rand(shape) + np.random.rand(shape)*1.0j).astype(dtype)
return np.random.rand(shape).astype(dtype)
```
I haven't tested this, and it may need some tidying up but it should construct the matrix without raising a warning. Another way (which may be better) would be to ignore the warning *inside* the function. You can do this using [numpy.testing.suppress_warnings](https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.suppress_warnings.html)

Reply:
@swallan I think Kai's suggestion to do this with a conditional is good. Let's do that.